File: Data Structures & Algorithms/subsets/submission-4.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        
        n = len(nums)
        output = []

        def dfs(index, curr_array):

            if index >= n: 
                output.append(curr_array[:])
                return
            
            # add to the curr array
            curr_array.append(nums[index])
            dfs(index + 1, curr_array)
            print(curr_array.pop())

            # NOT add to the curr array
            dfs(index + 1, curr_array)


        dfs(0, [])
        return output
        

File: Data Structures & Algorithms/daily-temperatures/submission-4.cpp
class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& temperatures) {
        std::vector<int> warmer_temp_appears(temperatures.size(), 0);

        std::stack<std::pair<int, int>> temp_and_index;

        // less than or equal to top of stack, push on stack

        // while the curr number is > top of stack, pop it and update the index in warmer_temp_appears

        for (int i{0}; i < temperatures.size(); i++){
            int curr_temp = temperatures[i];

            while(!temp_and_index.empty() && curr_temp > temp_and_index.top().first){
                int temp = temp_and_index.top().first;
                int index = temp_and_index.top().second;

                warmer_temp_appears[index] = i - index;

                // remove from stack
                temp_and_index.pop();
            }
            temp_and_index.push({curr_temp, i});
        }
        return warmer_temp_appears;
    }
};